app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt

```diff
@@ -4,51 +4,164 @@
+                                }
+                            )
+                        }
+                }
+
+                history = getHistoryAndSearchGroups(offset, numberOfItems)
+            } else {
+                history = historyStorage
+                    .getVisitsPaginated(
+                        offset,
+                        numberOfItems,
+                        excludeTypes = listOf(
+                            VisitType.NOT_A_VISIT,
+                            VisitType.DOWNLOAD,
+                            VisitType.REDIRECT_TEMPORARY,
+                            VisitType.RELOAD,
+                            VisitType.EMBED,
+                            VisitType.FRAMED_LINK,
+                            VisitType.REDIRECT_PERMANENT
+                        )
                     )
+                    .mapIndexed(transformVisitInfoToHistoryItem(offset.toInt()))
+            }
+
+            onComplete(history)
+        }
+    }
+
+    @Suppress("MagicNumber")
+    private suspend fun getHistoryAndSearchGroups(
+        offset: Long,
+        numberOfItems: Long,
+    ): List<History> {
+        val result = mutableListOf<History>()
+        val history: List<History.Regular> = historyStorage
+            .getVisitsPaginated(
+                offset,
+                numberOfItems,
+                excludeTypes = listOf(
+                    VisitType.NOT_A_VISIT,
+                    VisitType.DOWNLOAD,
+                    VisitType.REDIRECT_TEMPORARY,
+                    VisitType.RELOAD,
+                    VisitType.EMBED,
+                    VisitType.FRAMED_LINK,
+                    VisitType.REDIRECT_PERMANENT
                 )
+            )
+            .mapIndexed(transformVisitInfoToHistoryItem(offset.toInt()))
+
+        // History metadata items are recorded after their associated visited info, we add an
+        // additional buffer time to the most recent visit to account for a history group
+        // appearing as the most recent item.
+        val visitedAtBuffer = if (offset == 0L) 15000 else 0 /* 15 seconds in ms */
+
+        // Get the history groups that fits within the range of visited times in the current history
+        // items.
+        val historyGroupsInOffset = if (history.isNotEmpty()) {
+            historyGroups.filter {
+                history.last().visitedAt <= it.visitedAt &&
+                    it.visitedAt <= (history.first().visitedAt + visitedAtBuffer)
+            }
+        } else {
+            emptyList()
+        }
+        val historyMetadata = historyGroupsInOffset.flatMap { it.items }
 
-                onComplete(history)
+        result.addAll(history.filter { item -> historyMetadata.find { it.url == item.url } == null })
+
+        // Filter history metadata items with no view time.
+        result.addAll(
+            historyGroupsInOffset.map { group ->
+                group.copy(items = group.items.filter { it.totalViewTime > 0 })
```

Comment:
On home, we dedupe based on URLs, but I think it's fine to NOT do that here, because the user expectation must be to see all visits.

Reply:
Let's do that in a follow-up though because maybe we want some version of this deduping for regular history as well. It just seems kind of weird that we duplicate urls in a group.